### PR TITLE
[Wip] master stateless overload

### PR DESCRIPTION
**Pull request by yuit**
Update of the previous PR #12107

Fix: #11187 (see: tsxSpreadAttributesResolution1.tsx), #9703 (see : tsxStatelessFunctionComponentOverload6.tsx)

**Overview of the change**
- Treat `JSXAttributes` similar to `ObjectLiteralExpression`. We change the AST structure and store JSXAttributes as a AST Node instead of NodeArray
- Using overload resolution logic in choosing signature of stateless function component

**Note**
Currently when we see spread of any type, the entire attributes has any type which means that follow example is allowed even though we should give an error:

``` typescript
declare function MainBuntton(attr1: {to: string}): JSX.Element;
declare function MainBuntton(attr2: {onClick: ()=> void}): JSX.Element;
declare var obj3: any;
const b = <MainBuntton {...obj3} to={10} />;  // should error for mismatched type
```

### CODE REVIEW

**Comment by yuit** [comment]
@RyanCavanaugh has helped code-review this offline 🍰

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -1824,15 +1826,21 @@ namespace ts {
...
             write("/>");
         }
 
         function emitJsxOpeningElement(node: JsxOpeningElement) {
             write("<");
             emitJsxTagName(node.tagName);
-            writeIfAny(node.attributes, " ");
-            emitList(node, node.attributes, ListFormat.JsxElementAttributes);
+            writeIfAny(node.attributes.properties, " ");
+            // We are checking here so we won't re-enter the emiting pipeline and emit extra sourcemap
```
Comment (yuit) [inline src/compiler/emitter.ts]: I will confirm with Ron. My understanding is that we only want to call emit if the attributes are not empty (as this can cause extra sourcemap)

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -12256,41 +12481,12 @@ namespace ts {
...
-            // was spreaded in, though, assume that it provided all required properties
-            if (targetAttributesType && !sawSpreadedAny) {
-                const targetProperties = getPropertiesOfType(targetAttributesType);
-                for (const targetProperty of targetProperties) {
-                    if (!(targetProperty.flags & SymbolFlags.Optional) && !nameTable.get(targetProperty.name)) {
-                        error(node, Diagnostics.Property_0_is_missing_in_type_1, targetProperty.name, typeToString(targetAttributesType));
-                    }
-                }
-            }
+            checkJsxAttributesAssignableToTagNameAttributes(node);
```
Comment (sandersn) [inline src/compiler/checker.ts]: can you move the definition of this function closer to its call?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -12744,6 +12940,13 @@ namespace ts {
             let isDecorator: boolean;
             let spreadArgIndex = -1;
 
+            if (isJsxOpeningLikeElement(node)) {
+                // For JSX opening-like element, we will ignore regular arity check (which is what is done here).
```
Comment (sandersn) [inline src/compiler/checker.ts]: can you make this comment shorter? Basically just give a link to the function that does the work instead: "The arity check will be done in `checkApplicableSignatureForJsxOpeningLikeElement` instead".

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -12943,7 +13146,47 @@ namespace ts {
...
+            if (callIsIncomplete) {
+                return true;
+            }
+
+            const headMessage = Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1;
+            // Stateless function components can have maximum of three arguments: "props", "context", and "updater".
+            // However "context" and "updater" are implicit and can't be specify by users. Only the first parameter, props,
+            // can be specified by users through attributes property.
+            const paramType = getTypeAtPosition(signature, 0);
+            const argType = checkExpressionWithContextualType(node.attributes, paramType, /*contextualMapper*/ undefined);
```
Comment (sandersn) [inline src/compiler/checker.ts]: I think `attributesType` is a better name.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -13422,6 +13669,11 @@ namespace ts {
                 // in arguments too early. If possible, we'd like to only type them once we know the correct
                 // overload. However, this matters for the case where the call is correct. When the call is
                 // an error, we don't need to exclude any arguments, although it would cause no harm to do so.
+                if (isJsxOpeningOrSelfClosingElement) {
+                    // If there is not result, just return the last one we try as a candidate.
```
Comment (sandersn) [inline src/compiler/checker.ts]: 1. It looks like this `if` should go before the preceding comment block that talks about `excludeArgument`.
2. I would delete the first line of the comment because it just repeats what the code itself says.

File: src/services/symbolDisplay.ts [hunk 6]
```diff
@@ -114,23 +117,33 @@ namespace ts.SymbolDisplay {
...
                     const candidateSignatures: Signature[] = [];
-                    signature = typeChecker.getResolvedSignature(callExpression, candidateSignatures);
+                    signature = typeChecker.getResolvedSignature(callExpressionLike, candidateSignatures);
                     if (!signature && candidateSignatures.length) {
                         // Use the first candidate:
                         signature = candidateSignatures[0];
                     }
 
-                    const useConstructSignatures = callExpression.kind === SyntaxKind.NewExpression || callExpression.expression.kind === SyntaxKind.SuperKeyword;
+                    let useConstructSignatures = false;
```
Comment (sandersn) [inline src/services/symbolDisplay.ts]: 1. I liked the single-expression form better.
2. Did you add `isCallExpression` because of this change? super doesn't seem related to stateless JSX overload resolution.

File: src/services/goToDefinition.ts [hunk 7]
```diff
@@ -85,6 +85,39 @@ namespace ts.GoToDefinition {
...
+        if (isJsxOpeningLikeElement(node.parent)) {
+            // For JSX opening-like element, the tag can be resolved either as stateful component (e.g class) or stateless function component.
+            // Because if it is a stateless function component with an error while trying to resolve the signature, we don't want to return all
+            // possible overloads but just the first one.
+            // For example:
+            //      /*firstSource*/declare function MainButton(buttonProps: ButtonProps): JSX.Element;
+            //      /*secondSource*/declare function MainButton(linkProps: LinkProps): JSX.Element;
+            //      /*thirdSource*/declare function MainButton(props: ButtonProps | LinkProps): JSX.Element;
+            //      let opt = <Main/*firstTarget*/Button />;  // We get undefined for resolved signature indicating an error, then just return the first declaration
+            const {symbolName, symbolKind, containerName}  = getSymbolInfo(typeChecker, symbol, node);
```
Comment (yuit) [inline src/services/goToDefinition.ts]: Nop we should fix that

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -7645,16 +7662,24 @@ namespace ts {
...
-                                errorNode = prop.valueDeclaration;
-                                reportError(Diagnostics.Object_literal_may_only_specify_known_properties_and_0_does_not_exist_in_type_1,
-                                    symbolToString(prop), typeToString(target));
+                                if (isJsxAttributes(errorNode)) {
+                                    // JsxAttributes has an object-literal flag and undergo same type-assignablity check as normal object-literal.
+                                    // However, using an object-literal error message will be very confusing to the users so we give different a message.
+                                    reportError(Diagnostics.Property_0_does_not_exist_on_type_1, symbolToString(prop), typeToString(target));
+                                }
+                                else {
+                                    errorNode = prop.valueDeclaration;
```
Comment (yuit) [inline src/compiler/checker.ts]: Talk offline, the error message helps elaborate this further so it should be ok for now.

File: src/services/symbolDisplay.ts (hunk 6 above)
Comment (yuit) [inline src/services/symbolDisplay.ts]: Yep because callExpressionLike include JSXOpeningElement as well. I did fix it up to use single-expression

File: tests/cases/fourslash/tsxQuickInfo6.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+//// function Bar<T extends {prop: number}>(arg: T) {
+////     let a1 = <Compone/*1*/ntSpecific {...arg} ignore-prop="hi" />;  // U is number
+////     let a2 = <ComponentSpecific1 {...arg} ignore-prop={10} />;  // U is number
+////     let a3 = <Component/*2*/Specific {...arg} prop="hello" />;   // U is "hello"
+//// }
+
+verify.quickInfos({
+    1: "function ComponentSpecific<{}>(l: {\n    prop: {};\n}): any",
+    2: "function ComponentSpecific<{}>(l: {\n    prop: {};\n}): any"
```
Comment (sandersn) [inline tests/cases/fourslash/tsxQuickInfo6.ts]: Why is `U={}` and the return type `any` instead of `JSX.Element`? Is the lack of spread types causing errors in the code?

File: tests/cases/fourslash/tsxSignatureHelp1.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+//// }
+
+//// function _buildMainButton({ onClick, children, className }: ButtonProps): JSX.Element {
+////     return(<button className={className} onClick={onClick}>{ children || 'MAIN BUTTON'}</button>);  
+//// }
+
+//// export function MainButton(props: ButtonProps): JSX.Element {
+////     return this._buildMainButton(props);
+//// }
+//// let e1 = <MainButton/*1*/ /*2*/
```
Comment (sandersn) [inline tests/cases/fourslash/tsxSignatureHelp1.ts]: what is the difference between `/*1*/` and `/*2*/` here? They look like they should be the same, and in fact they are.

File: tests/cases/fourslash/tsxSignatureHelp2.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+//// export function MainButton(buttonProps: ButtonProps): JSX.Element;
+//// export function MainButton(linkProps: LinkProps): JSX.Element;
+//// export function MainButton(props: ButtonProps | LinkProps): JSX.Element {
+////     return this._buildMainButton(props);
+//// }
+//// let e1 = <MainButton/*1*/ /*2*/
+
+goTo.marker("1");
+// The second overload is chosen because the first doesn't get selected by choose overload as it has incorrect arity
```
Comment (sandersn) [inline tests/cases/fourslash/tsxSignatureHelp2.ts]: This comment seems wrong, because it looks like the first overload is chosen after all.

File: tests/baselines/reference/contextuallyTypedStringLiteralsInJsxAttributes02.errors.txt [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    export function MainButton(props: ButtonProps | LinkProps): JSX.Element {
+        const linkProps = props as LinkProps;
+        if(linkProps.goTo) {
+            return this._buildMainLink(props);
+        }
+    
+        return this._buildMainButton(props);
+    }
+    
+    const b0 = <MainButton {...{onClick: (k) => {console.log(k)}}} extra />;  // k has type any
```
Comment (yuit) [inline tests/baselines/reference/contextuallyTypedStringLiteralsInJsxAttributes02.errors.txt]: ah. I fix the contextual type and forgot to update the comment. Thanks

File: tests/baselines/reference/tsxStatelessFunctionComponents1.errors.txt [hunk 13]
```diff
@@ -1,7 +1,14 @@
...
-tests/cases/conformance/jsx/file.tsx(21,15): error TS2339: Property 'naaaaaaame' does not exist on type 'IntrinsicAttributes & { name?: string; }'.
+tests/cases/conformance/jsx/file.tsx(16,16): error TS2322: Type '{ naaame: "world"; }' is not assignable to type 'IntrinsicAttributes & { name: string; }'.
+  Property 'naaame' does not exist on type 'IntrinsicAttributes & { name: string; }'.
+tests/cases/conformance/jsx/file.tsx(24,15): error TS2322: Type '{ name: 42; }' is not assignable to type 'IntrinsicAttributes & { name?: string; }'.
+  Type '{ name: 42; }' is not assignable to type '{ name?: string; }'.
+    Types of property 'name' are incompatible.
+      Type '42' is not assignable to type 'string'.
+tests/cases/conformance/jsx/file.tsx(26,15): error TS2322: Type '{ naaaaaaame: "no"; }' is not assignable to type 'IntrinsicAttributes & { name?: string; }'.
+  Property 'naaaaaaame' does not exist on type 'IntrinsicAttributes & { name?: string; }'.
+tests/cases/conformance/jsx/file.tsx(31,23): error TS2322: Type '{}' is not assignable to type 'IntrinsicAttributes & { "prop-name": string; }'.
```
Comment (yuit) [inline tests/baselines/reference/tsxStatelessFunctionComponents1.errors.txt]: yep. So the "extra-prop-name" doesn't exist in attribute type so it doesn't participate in the type check and get filter out

File: tests/baselines/reference/tsxSpreadAttributesResolution6.errors.txt [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    
+    class TextComponent extends React.Component<TextProps, {}> {
+        render() {
+            return <span>Some Text..</span>;
+        }
+    }
+    
+    // Error
+    let x = <TextComponent editable={true} />
+             ~~~~~~~~~~~~~
```
Comment (yuit) [inline tests/baselines/reference/tsxSpreadAttributesResolution6.errors.txt]: ah this error is related to this bug #13526

File: tests/cases/fourslash/tsxSignatureHelp1.ts (hunk 10 above)
Comment (yuit) [inline tests/cases/fourslash/tsxSignatureHelp1.ts]: you are right. I fix that

File: tests/cases/fourslash/tsxSignatureHelp2.ts (hunk 11 above)
Comment (yuit) [inline tests/cases/fourslash/tsxSignatureHelp2.ts]: you are right. the comment is not. I must have forgot to update it when I was experimenting

File: tests/baselines/reference/tsxSpreadAttributesResolution7.symbols [hunk 15]
```diff
@@ -0,0 +1,62 @@
+=== tests/cases/conformance/jsx/file.tsx ===
```
Comment (yuit) [inline tests/baselines/reference/tsxSpreadAttributesResolution7.symbols]: errors.txt. I am surprised the test doesn't fail

File: tests/cases/fourslash/tsxQuickInfo6.ts (hunk 9 above)
Comment (yuit) [inline tests/cases/fourslash/tsxQuickInfo6.ts]: The reason it is "any" because we will fail to get JSX.Element (as there is no react library loaded here)....I believe that U={} is due to lack of spread types

File: tests/baselines/reference/tsxAttributeResolution3.errors.txt [hunk 16]
```diff
@@ -43,17 +49,19 @@ tests/cases/conformance/jsx/file.tsx(39,8): error TS2322: Type '32' is not assig
...
     <test1 x="ok" {...obj5} />
-                  ~~~~~~~~~
-!!! error TS2606: Property 'x' of JSX spread attribute is not assignable to target property.
-!!! error TS2606:   Type 'number' is not assignable to type 'string'.
+           ~~~~~~~~~~~~~~~~
+!!! error TS2322: Type '{ x: number; y: number; }' is not assignable to type 'Attribs1'.
+!!! error TS2322:   Types of property 'x' are incompatible.
+!!! error TS2322:     Type 'number' is not assignable to type 'string'.
     
     // OK
```
Comment (yuit) [inline tests/baselines/reference/tsxAttributeResolution3.errors.txt]: what do you mean?

Comment (sandersn) [inline tests/baselines/reference/tsxAttributeResolution3.errors.txt]: Sorry, I meant *should* be. Right the comment is `// OK` but `<test1 {...obj6} />` actually causes an error.

**Comment by yuit** [comment]
🔔 @RyanCavanaugh  if you would like to give another look :)

File: src/compiler/checker.ts [hunk 17]
```diff
@@ -590,7 +590,7 @@ namespace ts {
             return node.kind === SyntaxKind.SourceFile && !isExternalOrCommonJsModule(<SourceFile>node);
         }
 
-        function getSymbol(symbols: SymbolTable, name: string, meaning: SymbolFlags): Symbol {
+       function getSymbol(symbols: SymbolTable, name: string, meaning: SymbolFlags): Symbol {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Wrong indent?

File: src/compiler/checker.ts [hunk 18]
```diff
@@ -3362,6 +3362,11 @@ namespace ts {
                 const type = checkDeclarationInitializer(declaration);
                 return addOptionality(type, /*optional*/ declaration.questionToken && includeOptionality);
             }
+            else if (isJsxAttribute(declaration)) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: No need for `else` here

**Comment by yuit** [comment]
Talk with @RyanCavanaugh, we should have better name for this one `defaultTryGetJsxStatelessFunctionAttributesType`

File: src/compiler/checker.ts [hunk 19]
```diff
@@ -13023,7 +13227,47 @@ namespace ts {
...
+            // can be specified by users through attributes property.
+            const paramType = getTypeAtPosition(signature, 0);
+            const attributesType = checkExpressionWithContextualType(node.attributes, paramType, /*contextualMapper*/ undefined);
+            const argProperties = getPropertiesOfType(attributesType);
+            for (const arg of argProperties) {
+                if (!getPropertyOfType(paramType, arg.name) && isUnhyphenatedJsxName(arg.name)) {
+                    return false;
+                }
+            }
+            if (checkTypeRelatedTo(attributesType, paramType, relation, /*errorNode*/ undefined, headMessage)) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Can just return this value directly

File: src/compiler/checker.ts [hunk 20]
```diff
@@ -13105,8 +13349,11 @@ namespace ts {
                 // `getEffectiveArgumentCount` and `getEffectiveArgumentType` below.
                 return undefined;
             }
+            else if (isJsxOpeningLikeElement(node)) {
+                args = node.attributes.properties.length > 0 ? [node.attributes] : [];
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Intentional to use `[]` instead of `emptyArray` here?

File: src/compiler/checker.ts [hunk 21]
```diff
@@ -13900,6 +14152,69 @@ namespace ts {
...
+         * @return a resolved signature if we can find function matching function signature through resolve call or a first signature in the list of functions.
+         *         otherwise return undefined if tag-name of the opening-like element doesn't have call signatures
+         */
+        function resolveStatelessJsxOpeningLikeElement(openingLikeElement: JsxOpeningLikeElement, elementType: Type, candidatesOutArray: Signature[]): Signature {
+            if (elementType.flags & TypeFlags.Union) {
+                const types = (elementType as UnionType).types;
+                let result: Signature;
+                for (const type of types) {
+                    // This is mainly to fill in all the candidates if there is one.
+                    result = result && resolveStatelessJsxOpeningLikeElement(openingLikeElement, type, candidatesOutArray);
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Shouldn't this be `||` ? I would expect a failing test because of this

File: src/compiler/checker.ts (hunk 20 above)
Comment (yuit) [inline src/compiler/checker.ts]: no. Fix

File: src/compiler/checker.ts (hunk 21 above)
Comment (yuit) [inline src/compiler/checker.ts]: We don't have test around this surprisingly. I am adding one in
